akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

```diff
@@ -0,0 +1,170 @@
+        }
+      }
+
+      override def postStop(): Unit = {
+        if (!innerMatValue.isCompleted) {
+          innerMatValue.failure(new AbruptStageTerminationException(this))
+        }
+      }
+
+      object Initializing extends InHandler with OutHandler {
+        override def onPush(): Unit = {
+          throw new IllegalStateException("unexpected push during initialization")
+        }
+
+        var upstreamFailure = OptionVal.none[Throwable]
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          upstreamFailure = OptionVal.Some(ex)
+        }
+
+        override def onUpstreamFinish(): Unit = {
+          //will later be propagated to the materialized flow
+        }
+
+        var hasBeenPulled = false
+        override def onPull(): Unit = {
+          hasBeenPulled = true
+        }
+
+        var downstreamCause = OptionVal.none[Throwable]
+        override def onDownstreamFinish(cause: Throwable): Unit = {
+          downstreamCause = OptionVal.Some(cause)
+        }
+
+        def onFuture(futureRes: Try[Flow[In, Out, M]]) = futureRes match {
+          case Failure(exception) =>
+            setKeepGoing(false)
+            innerMatValue.failure(new NeverMaterializedException(exception))
+            failStage(exception)
+          case Success(flow) =>
+            //materialize flow, connect to inlets, feed with potential events and set handlers
+            connect(flow)
+            setKeepGoing(false)
+        }
+      }
+      def connect(flow: Flow[In, Out, M]): Unit = {
+        val subSource = new SubSourceOutlet[In](s"${FutureFlow.this}.subIn")
+        val subSink = new SubSinkInlet[Out](s"${FutureFlow.this}.subOut")
+
+        subSource.setHandler {
+          new OutHandler {
+            override def onPull(): Unit = if (!isClosed(in)) tryPull(in)
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
+              if (!isClosed(in)) {
+                cancel(in, cause)
+              }
+            }
+          }
+        }
+        subSink.setHandler {
+          new InHandler {
+            override def onPush(): Unit = {
+              push(out, subSink.grab())
+            }
+
+            override def onUpstreamFinish(): Unit = {
+              complete(out)
+            }
+
+            override def onUpstreamFailure(ex: Throwable): Unit = {
+              fail(out, ex)
+            }
+          }
+        }
+        Try {
+          Source.fromGraph(subSource.source).viaMat(flow)(Keep.right).to(subSink.sink).run()(subFusingMaterializer)
+        } match {
+          case Success(matVal) =>
+            innerMatValue.success(matVal)
+            Initializing.upstreamFailure match {
+              case OptionVal.Some(ex) =>
+                subSource.fail(ex)
+              case OptionVal.None =>
+                if (isClosed(in))
+                  subSource.complete()
+            }
+            Initializing.downstreamCause match {
```

Comment:
Better pass these into the def for more clarity.

Reply:
@jrudolph , can u please explain? do u mean explicitly passing the vars as method parameters to the connect method?
perhaps it makes more sense to make the connect method a member of Initializing?